inference-engine/tests/functional/inference_engine/ov_async_infer_queue.cpp

```diff
@@ -0,0 +1,259 @@
+#include <openvino/runtime/async_infer_queue.hpp>
+
+using namespace ::testing;
+using namespace std;
+
+std::shared_ptr<ov::Function> create_add_model(ov::Shape shape) {
+    auto arg0 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto arg1 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto add = make_shared<ov::op::v1::Add>(arg0, arg1);
+
+    return make_shared<ov::Function>(add, ov::ParameterVector{arg0, arg1});
+}
+
+std::shared_ptr<ov::Function> create_mul_model(ov::Shape shape) {
+    auto arg0 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto arg1 = make_shared<ov::op::v0::Parameter>(ov::element::f32, shape);
+    auto mul = make_shared<ov::op::v1::Multiply>(arg0, arg1);
+
+    return make_shared<ov::Function>(mul, ov::ParameterVector{arg0, arg1});
+}
+
+TEST(AsyncInferQueueOVTests, BasicFlowTest) {
+    ov::runtime::Core core;
+    ov::runtime::ExecutableNetwork net =
+        core.compile_model(create_add_model({8}), "CPU");
+
+    ov::runtime::AsyncInferQueue basic_queue(net);
+
+    size_t counter = 0;
+    std::vector<ov::Any> callback_data;
+
+    auto f = [&counter, &callback_data](std::exception_ptr e,
+                                        ov::runtime::InferRequest &request,
+                                        const ov::Any &data) {
+        counter += 1;
+        callback_data.push_back(data);
+    };
+
+    basic_queue.set_callback(f);
+
+    size_t num_of_runs = 4;
+
+    for (size_t i = 0; i < num_of_runs; i++) {
+        basic_queue.start_async(i);
+    }
+
+    basic_queue.wait_all();
+
+    std::vector<size_t> converted_data;
+    std::transform(callback_data.begin(), callback_data.end(),
+                   std::back_inserter(converted_data),
+                   [](ov::Any val) -> std::size_t { return val.as<size_t>(); });
+
+    ASSERT_EQ(counter, num_of_runs);
+    ASSERT_THAT(converted_data, UnorderedElementsAreArray({0, 1, 2, 3}));
+}
+
+TEST(AsyncInferQueueOVTests, AccessRequestInsideCallbackTest) {
+    size_t num_of_runs = 5;
+    ov::Shape input_shape{1};
+    std::vector<float> results;
+    std::vector<float> expected_results;
+
+    ov::runtime::Core core;
+    ov::runtime::ExecutableNetwork net =
+        core.compile_model(create_add_model(input_shape), "CPU");
+
+    ov::runtime::AsyncInferQueue basic_queue(net, 3);
+
+    auto f = [&results](std::exception_ptr e,
+                        ov::runtime::InferRequest &request, const ov::Any &data) {
+        float* data_ptr = request.get_output_tensor().data<float>();
+        results.push_back(data_ptr[0]);
+    };
+
+    basic_queue.set_callback(f);
+
+    for (size_t i = 0; i < num_of_runs; i++) {
+        std::vector<float> input_data(input_shape[0], i);
+
+        expected_results.push_back(i + i);
+
+        auto t = ov::runtime::Tensor(ov::element::f32, input_shape);
+        std::memcpy(t.data(), reinterpret_cast<void*>(&input_data[0]), t.get_byte_size());
```

Comment:
`std::copy_n(t.data<float>(), t.get_size(), input_data.begin());`

Reply:
Great point. Are there benefits of using `std::copy_n` others than throwing ExecutionPolicy and calling std::terminate? I am just curious.

Reply:
Some static analyzers do not like `memcpy` as it is `unsafe` as it does not have output bounds check. 
I don't know why `std::copy_n` is  interpreted as a more safe method ( maybe because we can enable safe iterators using compiler keys) 
And i gave incorrect suggestion. The code should be:
`std::copy_n(input_data.begin(), t.get_size(), t.data<float>());`
BTW the tensor can accept an external pointer as a parameter:
ov::runtime::Tensor(ov::element::f32, input_shape, input_data.data());`